test/inductor/test_cache.py

```diff
@@ -0,0 +1,151 @@
+    used_cache_keys.append(new_cache_key)
+    inductor_config.fx_graph_cache = new_cache_key
+
+
+class CacheTestCase(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        cls._saved_config = inductor_config.save_config()
+
+    def tearDown(self):
+        super().tearDown()
+        inductor_config.load_config(self._saved_config)
+
+    def setUp(self):
+        super().setUp()
+        # We don't want a fixed seed for this test to avoid cache collisions
+        set_rng_seed(int(time.time()))
+
+
+if HAS_CPU:
+
+    class TestCodeGenCache(CacheTestCase):
+        def test_cache_function_output(self):
+            activate_unique_cache()
+
+            @torch.compile
+            def foo(input, weight, bias):
+                return F.relu(F.linear(input, weight, bias))
+
+            input = torch.randn(64, 32)
+            weight = torch.randn(16, 32)
+            bias = torch.randn(16)
+
+            base_result = foo(input, weight, bias)
+            base_compilation_metrics = torch._dynamo.utils.compilation_time_metrics
+
+            # The compilation metrics of the first compilation should show codegen
+            self.assertIn("Scheduler.codegen", base_compilation_metrics)
+
+            # Important to reset so in-memory guards don't prevent compilation
+            torch._dynamo.reset()
+
+            cached_result = foo(input, weight, bias)
+            cached_compilation_metrics = torch._dynamo.utils.compilation_time_metrics
+
+            # The cached_result coming from a cached kernel should be the same as the original one
+            # The compilation metrics of the cached compilation should show no codegen
+            self.assertEqual(base_result, cached_result)
+            self.assertNotIn("Scheduler.codegen", cached_compilation_metrics)
+
+        def test_cache_module_output(self):
+            activate_unique_cache()
+
+            model = torch.compile(Foo())
+
+            input = torch.rand(15, 15)
+
+            base_result = model(input)
+            base_compilation_metrics = torch._dynamo.utils.compilation_time_metrics
+
+            # The compilation metrics of the first compilation should show codegen
+            self.assertIn("Scheduler.codegen", base_compilation_metrics)
+
+            # Important to reset so in-memory guards don't prevent compilation
+            torch._dynamo.reset()
+
+            cached_result = model(input)
+            cached_compilation_metrics = torch._dynamo.utils.compilation_time_metrics
+            # The cached_result coming from a cached kernel should be the same as the original one
+            # The compilation metrics of the cached compilation should show no codegen
+            self.assertEqual(base_result, cached_result)
+            self.assertNotIn("Scheduler.codegen", cached_compilation_metrics)
+
+
+if HAS_CUDA and not TEST_WITH_ASAN:
+
+    class TestCodeGenCacheCuda(CacheTestCase):
+        def test_cache_cudagraphs(self):
```

Comment:
I looked into this. This reason this fails is we create the `CompiledFxGraph`, set its `current_callable` to the [cudagraph'd function](https://github.com/pytorch/pytorch/blob/ea29116794f3beec8ac5a97277f13eb2ab17ad96/torch/_inductor/compile_fx.py#L413) and then store a reference to the compiled graph in memory `cls.cache[key] = compiled_graph`.

Later, when discard the cudagraph with `torch._dynamo.reset()`, that in memory reference is kept alive in `cls.cache`. Next time the key is hit, we return this now-invalid cudagraph function.

This fixes that error:

```
--- a/torch/_inductor/compile_fx.py
+++ b/torch/_inductor/compile_fx.py
@@ -410,8 +410,8 @@ def compile_fx_inner(
             ):
                 boxed_forward_device_index.set(next(iter(compiled_graph.device_idxs)))
 
-            compiled_graph.current_callable = cudagraphify(
-                compiled_graph.get_current_callable(),
+            current_callable = cudagraphify(
+                compiled_graph.compiled_artifact,
```

However, now you an `assertEqual` failure on the output. I don't think this is related to cudagraphs. If you remove `mode=reduce-overhead` from the test and force a File Lookup:
```
--- a/torch/_inductor/codecache.py
+++ b/torch/_inductor/codecache.py
@@ -454,6 +454,9 @@ class FxGraphCache:
                     with open(cg_path, "rb") as f:
                         compiled_graph = pickle.load(f)
 
+                # always force file lookup
+                return compiled_graph
+
                 cls.cache[key] = compiled_graph
```

you get the same error. I don't think we have any test coverage of file look up path.